usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2722,6 +2722,29 @@ typedef struct {
     enum cca_token_type keytype;
 } pkey_wrap_handler_data_t;
 
+/*
+ * On older kernels, the PKEY_KBLOB2PROTK3 ioctl is not yet available, which is
+ * no problem at runtime: protected key support is then just not available.
```

Comment:
> protected key support is then just not available.

Are you sure that you current code behaves so? If the ioctl fails, all ends with CKR_FUNCTION_FAILED and no secure key fallback is taken.

Also on each key usage it would try again and again to use the ioctl and fail again.
So you should detect during init if the kernel supports that ioctl or not.
Is this maybe done when getting the firmware MKVP ?

Reply:
> Is this maybe done when getting the firmware MKVP ?

Yes, routine ccatok_pkey_get_firmware_wkvp sets cca_data->pkey_wrap_supported = 1; if it could successfully determine the wkvp. This indicates if pkey support is available.